### DISCUSSION TITLE
tailwindcss -w doesn't rebuild CSS file, continuously appends new CSS styles, expected?!

### DISCUSSION THREAD

**Discussion by wgordon17** [tailwindlabs/tailwindcss#10186] [Help]
When running `npm run taildwindcss` locally with `-- -w`, it says that it's "Rebuildling...", but looking at the generated CSS file, it looks like it just keeps appending the new styles, perhaps with the _expectation_ that the new styles will supersede the old ones? This just creates confusion when looking in Chrome Devtools.

Is this the correct/expected behavior?!

**Comment by adamwathan** [comment]
New styles shouldn't be _appended_ but rather inserted in the stylesheet in the right location in a sort order that Tailwind maintains internally. Are you sure that hitting save always causes the new styles you've added to be literally appended?

Tailwind doesn't remove any old styles even if unused anymore for performance reasons, but of course a fresh one-off build (like you'd do for production) won't include any styles that accumulated while developing with watch mode.

**Reply by wgordon17** [reply]
That's what I would've thought too! I actually managed to get this reproducible: https://github.com/wgordon17/tailwind-reproducer

If you just run `npx tailwindcss -i style.css -o dist/style.css -w` and then make modifications to the `div` class, like

- `[&>div>label>span]:hidden`
- `[&>div>label>span]:inline`
- `first:[&>div>label>span]:hidden`
- `last:[&>div>label>span]:hidden`

It will just keep appending to the `dist/style.css` file, e.g.

```
.\[\&\>div\>label\>span\]\:inline>div>label>span {
  display: inline;
}

.\[\&\>div\>label\>span\]\:hidden>div>label>span {
  display: none;
}

.first\:\[\&\>div\>label\>span\]\:inline>div>label>span:first-child {
  display: inline;
}

.last\:\[\&\>div\>label\>span\]\:inline>div>label>span:last-child {
  display: inline;
}
```

This is definitely an _inconvenience_, but not the end of the world since, as you mention, I can just rerun the `npx tailwindcss` and it'll rebuild the file correctly. But it _is_ still annoying during development. Is this a bug? Any other suggestions on working around this? 🙏

**Reply by wgordon17** [reply]
This seems to be normal behavior even on _normal_ classes, just like `hidden` by itself ¯\\\_(ツ)_/¯ Maybe this is just an optimization, that it won't remove the classes while doing a watch, even if they're no longer being used?

**Reply by wgordon17** [reply]
> Tailwind doesn't remove any old styles even if unused anymore for performance reasons

Doh! Guess, I should've read this a little closer! So yes, clearly sounds like an optimization, as you mentioned. Although, I noticed on play.tailwindcss.com, the entire file was being recreated without this optimization. Is there an easy way to reproduce that behavior locally?

**Reply by adamwathan** [reply]
> Is there an easy way to reproduce that behavior locally?

Not currently, if you use watch mode we only add new styles and don't remove old ones.

To be clear you aren't seeing things be _appended_ right (which means always added to the very end of the file)? Adding the `hidden` class when the file already contains `hover:bg-black` for example shouldn't add `hidden` to the _end_, but to the utilities layer which is earlier in the file than `hover:bg-black`.

**Reply by wgordon17** [reply]
You're correct, my apologies for not paying enough attention. The styles are not being _appended_ by being added to the very end of the file they're being _inserted_ (typically near the previous style) in _addition_ to the previous style.

I _am_ noticing an issue with `media` queries, where it doesn't appear that they're being added in the proper _order_ when running `tailwindcss -w`. This issue fixes itself when rerunning the command, to force the file rebuild. Not sure if this is something known? ¯\\\_(ツ)_/¯

**Reply by wgordon17** [reply]
Btw, when I mention "proper order", I'm referring what I would _expect_ is that `media` queries are toward the bottom of the file, so that they have a higher precedence. 

The behavior that I'm seeing is that I have `hidden lg:block`, and even though I've viewing from a `lg` viewpoint, the element is still hidden. Again, if I rebuild the file, things sort themselves out, but if I'm rapidly developing and playing with breakpoints, I have seen it where the `media` query appears to be inserted somewhere incorrectly.
